osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs

```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                double p300 = logSum(p300Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p300Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p200 = logSum(p200Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p200Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p100 = logSum(p100Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p100Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p50 = logSum(p50Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p50Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+                double p0 = logSum(p0Note + Math.Log(attributes.NoteCount + attributes.HoldNoteCount), p0Tail + Math.Log(attributes.HoldNoteCount)) - Math.Log(totalJudgements);
+
+                double gradient = Math.Exp(
+                    (countPerfect * pMax
+                     + (countGreat + 0.5) * p300
+                     + countGood * p200
+                     + countOk * p100
+                     + countMeh * p50
+                     + countMiss * p0) / totalJudgements
+                );
+
+                return -gradient;
+            }
+
+            // Finding the minimum of the function returns the most likely deviation for the hit results.
+            return isLegacyScore ? FindMinimum.OfScalarFunction(legacyLikelihoodGradient, 10) : FindMinimum.OfScalarFunction(lazerLikelihoodGradient, 10);
+        }
+
+        private double[] getLegacyHitWindows(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] judgements = new double[5];
+
+            double overallDifficulty = attributes.OverallDifficulty;
+
+            if (attributes.Convert)
+                overallDifficulty = 10;
+
+            double windowMultiplier = 1;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            judgements[0] = Math.Floor(16 * windowMultiplier);
+            judgements[1] = Math.Floor((64 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[2] = Math.Floor((97 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[3] = Math.Floor((127 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[4] = Math.Floor((151 - 3 * overallDifficulty) * windowMultiplier);
+
+            return judgements;
+        }
+
+        private double[] getLazerHitWindows(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] hitWindows = new double[5];
+
+            var track = new TrackVirtual(10000);
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            double clockRate = track.Rate;
```

Comment:
I got your comment on a prior thread, I recommend to comment this code as it's quite vague. 
I'm uncertain if it's a standard/common way to yield the clock rate

Reply:
Added comments.